ipa-core/src/helpers/transport/stream/buffered.rs

```diff
@@ -0,0 +1,313 @@
+use std::{
+    mem,
+    num::NonZeroUsize,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use bytes::Bytes;
+use futures::Stream;
+use pin_project::pin_project;
+
+use crate::helpers::BytesStream;
+
+/// An adaptor to buffer items coming from the upstream
+/// [`BytesStream`](BytesStream) until the buffer is full, or the upstream is
+/// done. This may need to be used when writing into HTTP streams as Hyper
+/// does not provide any buffering functionality and we turn NODELAY on
+#[pin_project]
+pub struct BufferedBytesStream<S> {
```

Comment:
a few nits

To me `BufferedBytesStream` doesn't tell why this stream is special. I think the key thing is the poll with is being "chunked". I would rather call this `BufferedChunkedStream`

why not have the trait bound `S: BytesStream` here as well? Thinking more about documentation/readability

Reply:
Bounds need to be repeated everywhere if put on the struct - generally we try to avoid that if that's not necessary

Reply:
this does not attempt to chunk the inner stream, the whole purpose of it is to accumulate enough bytes (buffer) before sending them down for processing.

Reply:
But doesn't the output get chunked? This adapter is buffering and chunking... another name would be paginating.

I still think BufferedBytesStream is vague. My 2 cents.